File: 213.打家劫舍-ii.go
/*
 * @lc app=leetcode.cn id=213 lang=golang
 *
 * [213] 打家劫舍 II
 */

// @lc code=start
func rob(nums []int) int {
	n := len(nums)
	max := func(x,y int) int {
		if x > y {
			return x
		}
		return y
	}
	if n == 1 {
		return nums[0]
	}

	pre2 := nums[0]
	pre1 := max(pre2,nums[1])
	for i := 2 ; i < n-1 ; i++ {
		f := max(pre1,pre2+nums[i])
		pre2,pre1 = pre1,f
	}
	result := pre1
	if n != 2 {
		pre2 = nums[1]
		pre1 = max(pre2,nums[2])
		for i := 3 ; i < n ; i++ {
			f := max(pre1,pre2+nums[i])
			pre2,pre1 = pre1,f
		}
		result = max(result,pre1)
	}

	return result
}
// @lc code=end

File: 213.打家劫舍-ii.ts
/*
 * @lc app=leetcode.cn id=213 lang=typescript
 *
 * [213] 打家劫舍 II
 */

// @lc code=start
function rob(nums: number[]): number {
    if (nums.length <= 2) {
        return Math.max(...nums);
    }
    let dp1 = nums[0];
    let dp2 = nums[1];
    nums.forEach((num, index) => {
        if (index <= 1 || index === nums.length-1) {
            return
        }
        let dpi = Math.max(dp2, dp1 + num);
        dp1 = dp2;
        dp2 = dpi;
    })

    let res: number = dp2;
    if (nums.length > 2) {
        dp1 = nums[1];
        dp2 = nums[2];
        nums.forEach((num, index) => {
            if (index <= 2) {
                return
            }
            let dpi = Math.max(dp2, dp1 + num);
            dp1 = dp2;
            dp2 = dpi;
        })
    }
    res = Math.max(res, dp2);
    return res;
};
// @lc code=end

